beamer/agent/middleware.py

```diff
@@ -282,3 +311,31 @@ def rate_limiter(
 ) -> Callable[[RPCEndpoint, Any], RPCResponse]:
     state = _RateLimiterState()
     return functools.partial(_rate_limiter, make_request=make_request, w3=w3, state=state)
+
+
+def max_fee_setter(
+    make_request: _MakeRequest, _w3: Web3, chain_id: ChainId
+) -> Callable[[RPCEndpoint, Any], RPCResponse]:
+    def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+        if method != "eth_sendTransaction":
+            return make_request(method, params)
+        priority_fee_response = make_request(RPCEndpoint("eth_maxPriorityFeePerGas"), [])
+        if _result_ok(priority_fee_response):
+            priority_fee = priority_fee_response["result"]
+        else:
+            return priority_fee_response
+
+        latest_block = _BLOCK_STORAGE[chain_id].get_latest_block()
+        if latest_block is None:
+            latest_block_response = make_request("eth_getBlockByNumber", ["latest"])
+            if _result_ok(latest_block_response):
+                latest_block = latest_block_response
+            else:
+                return latest_block_response
+        base_fee = latest_block["result"]["baseFeePerGas"]
```

Comment:
nitpick: can we also use properties? `latest_block["result"].baseFeePerGas`? Would be nicer if we are consistent in the usage